tools/deploy.go

```diff
@@ -0,0 +1,711 @@
+		_, err := resourceClient.Update(ctx, obj, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("error updating %s %q: %w", obj.GetKind(), name, err)
+		}
+	} else {
+		return fmt.Errorf("error getting %s %q: %w", obj.GetKind(), name, err)
+	}
+
+	return nil
+}
+
+func execCommandToFile(cmd *exec.Cmd, outputPath string) error {
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("command execution failed: %w", err)
+	}
+
+	err = os.WriteFile(outputPath, output, 0600)
+	if err != nil {
+		return fmt.Errorf("writing command output to file failed: %w", err)
+	}
+
+	return nil
+}
+
+// deployWithKustomizeAndApply first generates the YAML configuration by running Kustomize build on the overlay directory,
+// then outputs the generated YAML to a temp file within the overlay dir, and then applies
+// the configuration to the Kubernetes cluster specified in the kubeconfig file.
+func deployWithKustomizeAndApply(kustomizePath, overlayPath string) error {
+	yamlOutputPath := filepath.Join(overlayPath, "output.yaml")
+
+	defer func() {
+		if removeErr := os.Remove(yamlOutputPath); removeErr != nil {
+			log.Printf("Warning: Failed to remove temporary YAML file %s: %v", yamlOutputPath, removeErr)
+		}
+	}()
+
+	// Generate YAML with kustomize
+	cmd := exec.Command(kustomizePath, "build", overlayPath)
+	err := execCommandToFile(cmd, yamlOutputPath)
+	if err != nil {
+		return fmt.Errorf("failed to generate YAML with kustomize: %v", err)
+	}
+
+	err = applyYAML(yamlOutputPath)
+	if err != nil {
+		return fmt.Errorf("failed to apply YAML: %v", err)
+	}
+
+	return nil
+}
+
+func copyFile(src, dst string) error {
+	input, err := os.ReadFile(src)
+	if err != nil {
+		return fmt.Errorf("failed to read source file %s: %v", src, err)
+	}
+
+	err = os.WriteFile(dst, input, 0600)
+	if err != nil {
+		return fmt.Errorf("failed to write destination file %s: %v", dst, err)
+	}
+
+	return nil
+}
+
+func deployBMO(tempBMOOverlay, scriptDir, kustomizePath string) error {
+	revertFunc, err := changeDirTemporarily(tempBMOOverlay)
+	if err != nil {
+		log.Fatalf("Failed to change directory %s: %v", tempBMOOverlay, err)
+	}
+	defer revertFunc()
+
+	// This is to keep the current behavior of using the ironic.env file for the configmap
+	ironicEnvSrc := filepath.Join(scriptDir, "config", "default", "ironic.env")
+	ironicEnvDst := filepath.Join(tempBMOOverlay, "ironic.env")
+	copyFile(ironicEnvSrc, ironicEnvDst)
+
+	execCommand(kustomizePath, "edit", "add", "configmap", "ironic", "--behavior=create", "--from-env-file=ironic.env")
+
+	err = deployWithKustomizeAndApply(kustomizePath, tempBMOOverlay)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func replaceAllPlaceholdersInFile(filePath, placeholder, newValue string) error {
+	input, err := os.ReadFile(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to read file %s: %v", filePath, err)
+	}
+
+	output := strings.ReplaceAll(string(input), placeholder, newValue)
```

Comment:
So if the file has `PLACEHOLDER` and `PLACEHOLDER_ARGS` and you replace `PLACEHOLDER` with `foo` then you end up with `foo` and `foo_ARGS`, and then you cannot replace `PLACEHOLDER_ARGS` later...

Reply:
This code is now gone. I am using `sed` with exec.Command instead.